Bluetooth: Mesh: Shell: Make prov features dynamic

Andrewpini opened the pull request:
Alters several provisioning related features to allow commands to be functional with application defined Provisioning properties & capabilities.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

alxelax commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -1498,3 +1511,45 @@ SHELL_STATIC_SUBCMD_SET_CREATE(mesh_cmds,
+		return 0;
+	}
+
+	app_input = prov->input;
```
Input action is hardcoded in shell but relaying on app assumes to be any.
The same for output.

Andrewpini replied:
I'm not sure I understand. Could you elaborate?

alxelax replied:
Take a look at hardcoded oob auth methods: https://github.com/Andrewpini/zephyr/blob/shell_prov_comp/subsys/bluetooth/mesh/shell/shell.c#L374-L378

alxelax reviewed (commented):
I do not like the idea to use part of the provisioning functions from the application and part from shell. It will cause some exotic combinations that we cannot predict how they work. It pushes us to implement universal OOB authentications for shell but actually, it doesn't have to.

I'd prefer to have only 2 options to use only app settings or to use only shell prov\capabilities.

PavelVPV commented:
> I do not like the idea to use part of the provisioning functions from the application and part from shell. It will cause some exotic combinations that we cannot predict how they work. It pushes us to implement universal OOB authentications for shell but actually, it doesn't have to.
> 
> I'd prefer to have only 2 options to use only app settings or to use only shell prov\capabilities.

Makes sense to me. But then I guess this API can be shared through the shell without any harm and manipulating the prov ctx:
https://github.com/zephyrproject-rtos/zephyr/blob/878edeb6ff190a713910a0e5cb9022704c5934a3/include/zephyr/bluetooth/mesh/main.h#L281-L384

It should also be possible to change UUID and static oob key for provisionee through shell.

Not sure if the shell prov ctx should be compiled out through a Kconfig option. It should be done at linkage I guess if `bt_mesh_shell_prov` is not used anywhere.

Andrewpini commented:
With the current changes it seems everything is still working (without the internal prov context) except the beacon listen command and the internal target utility feature. The interference with the prov context is reduced to the UUID and the static value.

alxelax commented on subsys/bluetooth/mesh/shell/shell.h:
```diff
@@ -0,0 +1,9 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
```
why is it required here?

Andrewpini replied:
Guess I forgot to remove it 👍

alxelax commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -720,6 +820,7 @@ static int cmd_beacon_listen(const struct shell *sh, size_t argc,
 
 	return 0;
 }
+#endif
```
```suggestion
#endif /* CONFIG_BT_MESH_SHELL_PROV_CTX_INSTANCE */
```

alxelax commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -601,6 +694,8 @@ static int cmd_rpl_clear(const struct shell *sh, size_t argc, char *argv[])
 }
 
+
+#if defined(CONFIG_BT_MESH_SHELL_PROV_CTX_INSTANCE)
```
Shouldn't `struct bt_mesh_prov bt_mesh_shell_prov` be wrapped up in this macro either?

Andrewpini replied:
Does it need to? Im not sure if extern variables will be compiled if they are not used anywhere 🤔

alxelax replied:
Hm. Is it extern? https://github.com/Andrewpini/zephyr/blob/shell_prov_comp/subsys/bluetooth/mesh/shell/shell.c#L441-L460

alxelax replied:
It is better to catch during the linking stage if you try to get access to a variable that is not used for provisioning. If it is kept as is you have a chance to use your own provisioning procedure but start beacon listening with a built-in one.

Andrewpini replied:
Seems extern: https://github.com/zephyrproject-rtos/zephyr/blob/ffcbc391efab079383ed7dc3561ead06b0107485/include/zephyr/bluetooth/mesh/shell.h#L46-L47

alxelax replied:
Yep, this allows it easy to build some samples like `mesh_shell` with `CONFIG_BT_MESH_SHELL_PROV_CTX_INSTANCE=n` and continue to use `bt_mesh_shell_prov`. This is an obvious misconfiguration. I'm pretty sur this should be fixed using condition compilation.

alxelax commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -361,33 +448,37 @@ struct bt_mesh_prov bt_mesh_shell_prov = {
+		prov->static_val_len = 0U;
 	} else {
-		bt_mesh_shell_prov.static_val_len = hex2bin(argv[1], strlen(argv[1]),
+		prov->static_val_len = hex2bin(argv[1], strlen(argv[1]),
```
Is it required `static_val` to be a static global variable?

Andrewpini replied:
You mean that we should not provide a way to set/alter the static value?

alxelax commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -361,33 +448,37 @@ struct bt_mesh_prov bt_mesh_shell_prov = {
 
 static int cmd_static_oob(const struct shell *sh, size_t argc, char *argv[])
 {
+	struct bt_mesh_prov *prov = (struct bt_mesh_prov *)bt_mesh_prov_get();
```
It is not clear to me if `CONFIG_BT_MESH_SHELL_PROV_CTX_INSTANCE` is set to `y` how static OOB value will be set there? 
Could you clarify that?

Andrewpini replied:
As I understand it the prov context pointer is passed from the app to `bt_mesh_init` which is then passed to  `bt_mesh_prov_init`. prov.c stores this pointer reference and when we call `bt_mesh_prov_get` in shell.c we get access to this pointer. The `cmd_static_oob` will be agnostic to the fact if we are using the internal prov context or not. We could add a "shorcut" for the instance where `CONFIG_BT_MESH_SHELL_PROV_CTX_INSTANCE ` is enabled so that we could use direct access to the internal context, but I don't see the point in doing that if we nevertheless want to access a prov context other than the internal context

Andrewpini commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -359,9 +390,8 @@ Configuration Client
 The Configuration Client model is an optional mesh subsystem that can be enabled through the :kconfig:option:`CONFIG_BT_MESH_CFG_CLI` configuration option. This is implemented as a separate module (``mesh models cfg``) inside the ``mesh models`` subcommand list. This module will work on any instance of the Configuration Client model if the mentioned shell configuration options is enabled, and as  ...
 
-The Configuration Client uses the general messages parameters set by ``mesh target dst`` and ``mesh target net`` to target specific nodes. When the Bluetooth mesh shell node is provisioned, the Configuration Client model targets itself by default. When another node has been provisioned by the Bluetooth mesh shell, the Configuration Client model targets the new node. In most common use-cases the C ...
+The Configuration Client uses the general messages parameters set by ``mesh target dst`` and ``mesh target net`` to target specific nodes. When the Bluetooth mesh shell node is provisioned, given that the :kconfig:option:`BT_MESH_SHELL_PROV_CTX_INSTANCE` is enabled with the proper application setup, the Configuration Client model targets itself by default. Similarly, when another node has been pr ...
```
That the user has added the internal pov context during init

mia-ko replied:
If you are asking this question, Alex, what are the chances that this will be clear to the user? Maybe we should rather add this information then :)

Andrewpini commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -318,10 +309,50 @@ To allow a device to provision devices over GATT, the :kconfig:option:`CONFIG_BT
+
+``mesh prov auth-method static <Value>``
+----------------------------------------
+	From the provisioner device, instruct the unprovisioned device to use static OOB authentication, and use the given static authentication value when provisioning.
```
Maybe? This is basically a copy/paste from the main.h docs

Andrewpini commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -400,28 +491,30 @@ static int cmd_uuid(const struct shell *sh, size_t argc, char *argv[])
 
-	memcpy(dev_uuid, uuid, len);
-	(void)memset(dev_uuid + len, 0, sizeof(dev_uuid) - len);
+	(void)memset(uuid + len, 0, sizeof(uuid) - len);
```
🤷‍♂️

alxelax replied:
seems kind of useless

mia-ko commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -318,10 +309,50 @@ To allow a device to provision devices over GATT, the :kconfig:option:`CONFIG_BT
 ------------------------------------------
 
-	Enable or disable printing of incoming unprovisioned beacons. Allows a provisioner device to detect nearby unprovisioned devices and provision them.
+	Enable or disable printing of incoming unprovisioned beacons. Allows a provisioner device to detect nearby unprovisioned devices and provision them. To enable this command the :kconfig:option:`BT_MESH_SHELL_PROV_CTX_INSTANCE` must be enabled.
```
```suggestion
	Enable or disable printing of incoming unprovisioned beacons. Allows a provisioner device to detect nearby unprovisioned devices and provision them. To enable this command, the :kconfig:option:`BT_MESH_SHELL_PROV_CTX_INSTANCE` option must be enabled.
```

mia-ko commented on doc/connectivity/bluetooth/api/mesh/shell.rst:
```diff
@@ -359,9 +390,8 @@ Configuration Client
 The Configuration Client model is an optional mesh subsystem that can be enabled through the :kconfig:option:`CONFIG_BT_MESH_CFG_CLI` configuration option. This is implemented as a separate module (``mesh models cfg``) inside the ``mesh models`` subcommand list. This module will work on any instance of the Configuration Client model if the mentioned shell configuration options is enabled, and as  ...
 
-The Configuration Client uses the general messages parameters set by ``mesh target dst`` and ``mesh target net`` to target specific nodes. When the Bluetooth mesh shell node is provisioned, the Configuration Client model targets itself by default. When another node has been provisioned by the Bluetooth mesh shell, the Configuration Client model targets the new node. In most common use-cases the C ...
+The Configuration Client uses the general messages parameters set by ``mesh target dst`` and ``mesh target net`` to target specific nodes. When the Bluetooth mesh shell node is provisioned, given that the :kconfig:option:`BT_MESH_SHELL_PROV_CTX_INSTANCE` is enabled with the proper application setup, the Configuration Client model targets itself by default. Similarly, when another node has been pr ...
```
```suggestion
The Configuration Client uses general message parameters set by ``mesh target dst`` and ``mesh target net`` to target specific nodes. When the Bluetooth mesh shell node is provisioned, given that the :kconfig:option:`BT_MESH_SHELL_PROV_CTX_INSTANCE` option is enabled with the proper application setup, the Configuration Client model targets itself by default. Similarly, when another node has been provisioned by the Bluetooth mesh shell, the Configuration Client model targets the new node. In most common use-cases, the Configuration Client is depending on the provisioning features and the Configuration database to be fully functional. The Configuration Client always sends messages using the Device key bound to the destination address, so it will only be able to configure itself and the mesh nodes it provisioned. The following steps are an example of how you can set up a device to start using the Configuration Client commands:
```

mia-ko reviewed (changes requested):
Just a few typos, otherwise LGTM :)

PavelVPV commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -400,28 +491,30 @@ static int cmd_uuid(const struct shell *sh, size_t argc, char *argv[])
-	memcpy(dev_uuid, uuid, len);
-	(void)memset(dev_uuid + len, 0, sizeof(dev_uuid) - len);
+	(void)memset(uuid + len, 0, sizeof(uuid) - len);
+	memcpy((uint8_t *)prov->uuid, uuid, sizeof(uuid));
```
The same as for `static_val`. It doesn't seem to be legal because uuid defined as const-qualifier.

PavelVPV commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -1299,10 +1399,7 @@ static int cmd_dst(const struct shell *sh, size_t argc, char *argv[])
-			    bt_mesh_shell_target_ctx.dst == bt_mesh_shell_target_ctx.local
-				    ? " (local)"
-				    : "");
+		shell_print(sh, "Destination address: 0x%04x", bt_mesh_shell_target_ctx.dst);
```
Why was it changed?

Andrewpini replied:
It is not valid if the internal prov context is not used. It could be conditional print for when CONFIG_BT_MESH_SHELL_PROV_CTX_INSTANCE is enabled, but I choose the simple fix here.

Andrewpini replied:
I think by using `bt_mesh_primary_addr()` we can get rid of the `local` parameter all togheter.

PavelVPV reviewed (changes requested):
I definitely missed that `bt_mesh_prov_get` returns `bt_mesh_prov` with const-qualifier and const `bt_mesh_prov` is passed to `bt_mesh_init`. This means that we can't change anything in the prov ctx through `bt_mesh_prov_get` and the commands that were doing changing the prov ctx content (like `uuid` and `static-oob`) should be used with the shell prov ctx only.

Andrewpini commented:
> I definitely missed that `bt_mesh_prov_get` returns `bt_mesh_prov` with const-qualifier and const `bt_mesh_prov` is passed to `bt_mesh_init`. This means that we can't change anything in the prov ctx through `bt_mesh_prov_get` and the commands that were doing changing the prov ctx content (like `uuid` and `static-oob`) should be used with the shell prov ctx only.

Ok, I feelt that I was playing a bit with fire here. Will rectify it 👍

Andrewpini committed:
Bluetooth: Mesh: Shell: Make prov features dynamic

Alters several provisioning related features to allow commands to be
functional with application defined Provisioning properties &
capabilities.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>
